impress-org/givewp
# New: Preload Donation Forms admin page data

## Description


Link preloading can be used to speed up page loads by making requests for resources early. Since the Donation Forms page JS first fully loads and then makes a fetch request, I anticipated there could be a significant page load speed increase by preloading its data.

First, I tried preloading the fetch using `<link rel="preload">`, but that doesn't allow you to set necessary headers and didn't work with WP REST due to caching. Next, I tried using [@wordpress/api-fetch](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-api-fetch/) along with `usePreloadingMiddleware`, but that ended up being a poorly documented path that led to a brittle solution.

So I settled on this solution: make a REST request manually before enqueuing scripts, then include the resulting data in the global along with the WP REST API details. This can be backported to the Logs admin page or any other page that makes a predictable WP REST API request on load. With a piece of added middleware, this works cleanly with SWR, regardless of fetcher function used. On pages that don't use SWR, it's as simple as not making a fetch to the WP REST API on page load and using the preloaded data as the initial state. 

This PR speeds up the load process of the Donation Forms page, making it feel more snappy. The results are most pronounced when on faster connections, because the biggest savings come from skipping the fixed overhead of a single HTTP/S request.

## Affects


N/A

## Visuals



https://user-images.githubusercontent.com/7819986/158235079-0ee0c3ee-e4e5-435e-8489-f00a7f205fdd.mp4
Left: After this PR
Right: Before this PR

## Testing Instructions


Install GiveWP
Navigate to Donations -> All Forms
Make sure you have at least 11 donation forms to test pagination; if not, you can duplicate forms until you have 11
Verify that the page loads correctly
Verify that changing pages works correctly, loading new data
Verify that searching forms by title/ID works correctly

## Pre-review Checklist
